File: examples/typescript/basic_usage.ts
/**
 * Basic MemoClaw usage — store, recall, iterate, and graph traversal.
 */
import { MemoClawClient } from '@memoclaw/sdk';

const client = new MemoClawClient({ wallet: '0xYourWalletAddress' });

// Store a memory
const stored = await client.store({
  content: 'User prefers TypeScript over JavaScript',
  importance: 0.8,
  metadata: { tags: ['preferences', 'language'] },
  memory_type: 'preference',
  namespace: 'user-prefs',
});
console.log(`Stored: ${stored.id}`);

// Recall by semantic search
const recalled = await client.recall({ query: 'What programming language?', limit: 5 });
for (const mem of recalled.memories) {
  console.log(`  [${mem.similarity.toFixed(2)}] ${mem.content}`);
}

// Iterate over ALL memories (automatic pagination)
for await (const memory of client.listAll({ namespace: 'user-prefs', batchSize: 25 })) {
  console.log(`  ${memory.id}: ${memory.content.slice(0, 60)}...`);
}

// Graph traversal
const graph = await client.getMemoryGraph(stored.id, 2);
for (const [mid, relations] of graph) {
  console.log(`  ${mid}: ${relations.length} relations`);
}

// Middleware hooks (fluent API)
const loggedClient = new MemoClawClient({ wallet: '0x...' })
  .onBeforeRequest((method, path) => { console.log(`→ ${method} ${path}`); })
  .onAfterResponse((method, path) => { console.log(`← ${method} ${path} OK`); })
  .onError((method, path, err) => { console.error(`✗ ${method} ${path}: ${err.message}`); });

await loggedClient.store({ content: 'With logging!' });

File: examples/python/basic_usage.py
"""Basic MemoClaw usage — store, recall, and iterate memories."""

from memoclaw import MemoClaw

# Initialize (reads MEMOCLAW_PRIVATE_KEY from env)
client = MemoClaw()

# Store a memory
result = client.store(
    "User prefers dark mode and vim keybindings",
    tags=["preferences", "editor"],
    importance=0.8,
    memory_type="preference",
)
print(f"Stored: {result.id} (tokens: {result.tokens_used})")

# Semantic recall
matches = client.recall("what editor settings does the user like?", limit=5)
for mem in matches.memories:
    print(f"  [{mem.similarity:.2f}] {mem.content}")

# Batch store
batch = client.store_batch([
    {"content": "User's timezone is UTC-5", "memory_type": "preference"},
    {"content": "Project deadline is March 2026", "memory_type": "project"},
    {"content": "User prefers Python over JavaScript", "memory_type": "preference"},
])
print(f"Batch stored {batch.count} memories")

# Iterate all memories (auto-paginates)
print("\nAll memories:")
for mem in client.iter_memories(batch_size=10):
    print(f"  [{mem.memory_type}] {mem.content[:60]}")

client.close()
